### PR TITLE
Fix: Invalid JSON for search_after or …before should cause 4xx (#4996)

### LINKED ISSUE #4996: Invalid JSON for search_after or …before should cause 4xx
The 2/2/2023 Invicti scan repored both a `Stack Trace Disclosure (Python)` and `Internal Server Error` (HTTP status 500) finding as a result of requesting a URL with incomplete JSON specified as the `search_after` parameter.

Request:
```
GET /index/bundles?catalog=dcp2&filters=%7B%7D&order=desc&search_after=%2527&size=10&sort=bundleVersion HTTP/1.1
Host: service.dev.singlecell.gi.ucsc.edu
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
Accept-Encoding: gzip, deflate
Accept-Language: en-us,en;q=0.5
Cache-Control: no-cache
Referer: https://service.dev.singlecell.gi.ucsc.edu/index/bundles
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.0 Safari/537.36
```

Response:
```
HTTP/1.1 500 Internal Server Error
Access-Control-Allow-Headers: Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key
X-Amz-Cf-Id: yljUg4uYRSb6npAuq4HeXGYYmCDOI2gNSKkSzWv0yIbFUf3-HRmPFA==
Connection: keep-alive
X-Amzn-Trace-Id: Root=1-63dc8df6-04db7de25839062032153ee5;Sampled=0
x-amz-apigw-id: fvseiHaKIAMFg7g=
Content-Length: 798
X-Cache: Error from cloudfront
Access-Control-Allow-Origin: *
Content-Type: text/plain
X-Amz-Cf-Pop: SFO20-C1
Date: Fri, 03 Feb 2023 04:30:46 GMT
Via: 1.1 1b76276e9fd9c7bd637ed463cc0a5976.cloudfront.net (CloudFront)
x-amzn-RequestId: b15a73ac-60c0-4d34-a657-8d2dd7079d83

Traceback (most recent call last):
File "/var/task/chalice/app.py", line 1913, in _get_view_function_response
response = view_function(**function_args)
File "/var/task/app.py", line 1219, in repository_search
pagination=app.get_pagination(entity_type),
File "/var/task/app.py", line 351, in get_pagination
sa = tuple(json.loads(sa))
File "/var/lang/lib/python3.9/json/__init__.py", line 346, in loads
return _default_decoder.decode(s)
File "/var/lang/lib/python3.9/json/decoder.py", line 337, in decode
obj, end = self.raw_decode(s, idx=_w(s, 0).end())
File "/var/lang/lib/python3.9/json/decoder.py", line 355, in raw_decode
raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

----

- [ ] Security design review completed; the Resolution of this issue does **not** …
  - [ ] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with 
  - [ ] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [ ] … affect the permissions of external users like access to
    - TDR snapshots
  - [ ] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [ ] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [ ] … affect monitoring of the system
  - [ ] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [ ] … add an interface that exposes sensitive or confidential data at the security boundary
  - [ ] … affect the encryption of data at rest
  - [ ] … require persistence of sensitive or confidential data that might require encryption at rest
  - [ ] … require unencrypted transmission of data within the security boundary
  - [ ] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [ ] Documentation on any unchecked boxes is provided in comments below

### PR DESCRIPTION
**Pull request by dsotirho-ucsc**
Connected issues: #4996

### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references all connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] For each connected issue, there is at least one commit whose title references that issue
- [x] PR is connected to all connected issues via ZenHub
- [x] PR description links to connected issues
- [x] Added `partial` label to PR <sub>or this PR completely resolves all connected issues</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title

### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR <sub>or this PR does not require reindexing</sub>
- [x] Added `a` (compatible changes) or `A` (incompatible ones) tag to commit title <sub>or this PR does not modify the Azul service API</sub>
- [x] Added `API` label to connected issues <sub>or this PR does not modify the Azul service API</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR <sub>or this PR does not touch requirements*.txt</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>

### Peer reviewer (after requesting changes)

Uncheck the *Author (before every review)* checklists.

### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *before every review* checklists. Update the `N reviews` label.

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Pushed PR branch to GitLab `dev` and added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved connected issues to Merged column in ZenHub
- [x] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [x] Shortened the PR chain <sub>or this PR is not labeled `base`</sub>
- [x] Pushed merge commit to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is*
pushed determines this checklist item.

### Operator (reindex)

- [x] Deleted unreferenced indices in `dev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvildev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `dev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvildev` deployment <sub>or this PR does not require reindexing</sub>

### Operator

- [x] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Review by nadove-ucsc — CHANGES_REQUESTED** [review]
Consider consolidating with the filters validation:

```diff
Subject: [PATCH] fixup!
---
Index: lambdas/service/app.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/lambdas/service/app.py b/lambdas/service/app.py
--- a/lambdas/service/app.py	(revision 335e08fe20bb3a94f51f065271bc3b13663fb7dd)
+++ b/lambdas/service/app.py	(date 1682623858598)
@@ -698,10 +698,7 @@
 
 
 def validate_filters(filters):
-    try:
-        filters = json.loads(filters)
-    except Exception:
-        raise BRE('The `filters` parameter is not valid JSON')
+    filters = validate_json(filters)
     if type(filters) is not dict:
         raise BRE('The `filters` parameter must be a dictionary')
     field_types = app.repository_controller.field_types(app.catalog)
@@ -783,9 +780,9 @@
         raise BRE(f'Unknown order `{order}`. Must be one of {supported_orders}')
 
 
-def validate_json(json_: str):
+def validate_json(json_: str) -> MutableJSON:
     try:
-        json.loads(json_)
+        return json.loads(json_)
     except json.decoder.JSONDecodeError:
         raise BRE(f'Invalid JSON string: {json_!r}')
 
```

**Comment by dsotirho-ucsc** [comment]
> Consider consolidating with the filters validation:

Implemented your suggestion with some tweaks to preserve the existing error message:
```
The `filters` parameter is not valid JSON
```
and produce a similar error message if `search_after` or `search_before` contain invalid JSON.

**Review by nadove-ucsc — CHANGES_REQUESTED** [review]
Nice work with the error messages. I think the test can be written more succinctly: 

```diff
Subject: [PATCH] fixup!
---
Index: test/service/test_response.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/service/test_response.py b/test/service/test_response.py
--- a/test/service/test_response.py	(revision 78588bade4a710e52eb19af4b97c52aef4383feb)
+++ b/test/service/test_response.py	(date 1682710468719)
@@ -2070,18 +2070,15 @@
             'search_before': response_json['pagination']['previous'],
             'search_after': response_json['pagination']['next']
         }
-        for pagination_key, url in test_cases.items():
+        for pagination_key, good_url in test_cases.items():
             with self.subTest(pagination_key=pagination_key):
-                query_params = parse_url_qs(url)
-                url_ = furl(url)
                 # Verify URL works before modifying
-                good_url = furl(origin=url_.origin, path=url_.path, query=query_params)
-                response = requests.get(str(good_url))
+                response = requests.get(good_url)
                 response.raise_for_status()
                 # Modify search_… param in URL and verify expected error occurs
-                self.assertIn('"', query_params[pagination_key])
-                query_params[pagination_key] = query_params[pagination_key].replace('"', '')
-                bad_url = furl(origin=url_.origin, path=url_.path, query=query_params)
+                bad_url = furl(good_url)
+                self.assertIn('"', bad_url.args[pagination_key])
+                bad_url.args[pagination_key] = bad_url.args[pagination_key].replace('"', '')
                 response = requests.get(str(bad_url))
                 error_msg = f'The `{pagination_key}` parameter is not valid JSON'
                 expected_text = f'{{"Code":"BadRequestError","Message":"{error_msg}"}}'
```

**Review by nadove-ucsc — DISMISSED** [review]

File: lambdas/service/app.py [hunk 1]
```diff
@@ -653,9 +653,9 @@ def validate_repository_search(params, **validators):
         'catalog': validate_catalog,
         'filters': validate_filters,
         'order': validate_order,
-        'search_after': str,
+        'search_after': validate_search_after,
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]:
```suggestion
        'search_after': partial(validate_json_param, 'search_before')
```

using `functools.partial`

File: lambdas/service/app.py [hunk 2]
```diff
@@ -783,6 +780,21 @@ def validate_order(order: str):
...
 
+def validate_search_before(search_before: str):
+    validate_json_param('search_before', search_before)
+
+
+def validate_search_after(search_after: str):
+    validate_json_param('search_after', search_after)
+
+
+def validate_json_param(name: str, json_: str) -> MutableJSON:
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]:
```suggestion
def validate_json_param(name: str, value: str) -> MutableJSON:
```

Underscore should be a last resort.

File: lambdas/service/app.py [hunk 3]
```diff
@@ -783,6 +780,21 @@ def validate_order(order: str):
...
+    try:
+        return json.loads(json_)
+    except json.decoder.JSONDecodeError:
+        raise BRE(f'The `{name}` parameter is not valid JSON')
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]:
```suggestion
        raise BRE(f'The {name!r} parameter is not valid JSON')
```

We're moving away from backquotes and there is precedent in this module that shows that we already started, so I'm OK with having a mix of single quotes and backquotes.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
PR and commit title should start with `Fix: `. I unchecked the corresponding CL item. Please do a split commit and push separately. And please squash out the fixup! commits.

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**